include/mppic/parameters_handler.hpp

```diff
@@ -0,0 +1,173 @@
+{
+  Dynamic,
+  Static
+};
+
+class ParametersHandler
+{
+public:
+  using get_param_func_t = void (const rclcpp::Parameter & param);
+  using post_callback_t = void ();
+  using pre_callback_t = void ();
+
+  ParametersHandler() = default;
+  ParametersHandler(const rclcpp_lifecycle::LifecycleNode::WeakPtr & parent)
+  {
+    node_ = parent;
+    auto node = node_.lock();
+    logger_ = node->get_logger();
+  }
+
+  void start()
+  {
+    auto node = node_.lock();
+    on_set_param_handler_ = node->add_on_set_parameters_callback(
+      std::bind(
+        &ParametersHandler::dynamicParametersCallback, this, std::placeholders::_1));
+  }
+
+  rcl_interfaces::msg::SetParametersResult dynamicParametersCallback(
+    std::vector<rclcpp::Parameter> parameters)
+  {
+    rcl_interfaces::msg::SetParametersResult result;
+
+    std::lock_guard<std::mutex> lock(parameters_change_mutex_);
+
+    for (auto & pre_cb : pre_callbacks_) {
+      pre_cb();
+    }
+
+    for (auto & param : parameters) {
+      const std::string & param_name = param.get_name();
+
+      if (auto callback = get_param_callbacks_.find(param_name);
+        callback != get_param_callbacks_.end())
+      {
+        RCLCPP_INFO(logger_, "Parameter %s found", param_name.c_str());
+        callback->second(param);
+      } else {
+        RCLCPP_WARN(logger_, "Parameter %s not found", param_name.c_str());
+      }
+    }
+
+    for (auto & post_cb : post_callbacks_) {
+      post_cb();
+    }
+
+    result.successful = true;
+    return result;
+  }
+
+  std::mutex * getLock()
+  {
+    return &parameters_change_mutex_;
+  }
+
+
+  template<typename T>
+  void addPostCallback(T && callback)
+  {
+    post_callbacks_.push_back(std::move(callback));
+  }
+
+  template<typename T>
+  void addPreCallback(T && callback)
+  {
+    pre_callbacks_.push_back(std::move(callback));
+  }
+
+  auto getParamGetter(std::string ns)
+  {
+    return [this, ns = std::move(ns)](auto & setting, std::string name, auto default_value,
+             ParameterType param_type = ParameterType::Dynamic) {
+             getParam(
+               setting, ns == "" ? name : ns + "." + name, std::move(
+                 default_value), param_type);
+           };
+  }
+
+  template<typename SettingT, typename ParamT>
+  void getParam(
+    SettingT & setting, std::string name, ParamT default_value,
+    ParameterType param_type = ParameterType::Dynamic)
+  {
+    auto node = node_.lock();
+
+    nav2_util::declare_parameter_if_not_declared(
+      node, name, rclcpp::ParameterValue(default_value));
+
+    setStaticParam<ParamT>(node, setting, name);
+
+    if (param_type == ParameterType::Dynamic) {
+      setDynamicParamCallback(setting, name);
+    }
+  }
+
+  template<typename ParamT, typename SettingT, typename NodeT>
+  void setStaticParam(NodeT node, SettingT & setting, std::string name)
+  {
+    ParamT param_in;
+    node->get_parameter(name, param_in);
+    setting = static_cast<SettingT>(param_in);
```

Comment:
I don't think this should ever be necessary, the ROS parameter system will throw an exception if the parameter declared type is different from the parameter set type in a parameter file. Ergo, if `get_parameter` doesn't throw an exception, we konw that the returned `param_in` type is the same as the declared setting type

Reply:
Setting here is actual parameter inside optimizer. So it could be unsigned int, but ros only allows int's so it must be casted